### PR TITLE
FEATURE: improves random assign automation

### PR DESCRIPTION
**Pull request by jjaffeux**
- Makes a best attempt at being random and assigning people who haven’t been assigned for a long time
- Uses timezones and holidays
- Allows to define a minimum delay between assignments
- Creates a post if no one is available

### CODE REVIEW

**Comment by jjaffeux** [comment]
@SamSaffron It's sadly very hard to test one plugin behavior relying on another plugin. Any idea on this?

File: lib/random_assign_utils.rb [hunk 1]
```diff
@@ -0,0 +1,37 @@
+# frozen_string_literal: true
+
+class RandomAssignUtils
+  def self.user_tzinfo(user_id)
```
Comment (eviltrout) [inline lib/random_assign_utils.rb:4]: I wonder if this method would be useful to have in core? It's a fair amount of logic and seems reusable.

File: plugin.rb [hunk 2]
```diff
@@ -567,14 +570,74 @@ def self.enqueue_assign_hooks(event, payload)
...
-        next unless topic = Topic.find_by(id: topic_id)
+        min_hours = (fields.dig('minimum_time_between_assignments', 'text') || 12).to_i
+        if TopicCustomField
+            .where(name: 'assigned_to_id', topic_id: topic_id)
+            .where('created_at < ?', min_hours.hours.ago)
+            .exists?
+          next
+        end
+
+        users_on_holiday = User
```
Comment (eviltrout) [inline plugin.rb:587]: I recommend using a `Set.new` for this. It'll make the `include?` below significantly faster if there are many users.

File: lib/random_assign_utils.rb (hunk 1 above)
Comment (jjaffeux) [reply lib/random_assign_utils.rb:4]: Sure, at the moment it's in dev-additions and I ported it over here.

Where would you see it defined?

```
User.tzinfo(user_id)
UserOption.tzinfo(user_id)
```

?

Comment (eviltrout) [reply lib/random_assign_utils.rb:4]: I'd say `UserOption` to keep `User` shorter.

**Comment by SamSaffron** [comment]
I think given this design we have I think CI is going to have to always include Discourse Automation, when testing any plugin. 

We should do a TODO to discuss this problem. Maybe we need metadata on plugins that specifies which plugins to load when running tests.

**Review by eviltrout — APPROVED** [review]

**Pull request merged**